runtime/resolvers/metricsview_cache_key.go

```diff
@@ -0,0 +1,159 @@
+package resolvers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/mitchellh/hashstructure/v2"
+	runtimev1 "github.com/rilldata/rill/proto/gen/rill/runtime/v1"
+	"github.com/rilldata/rill/runtime"
+	"github.com/rilldata/rill/runtime/metricsview"
+	"github.com/rilldata/rill/runtime/pkg/mapstructureutil"
+)
+
+var errCachingDisabled = errors.New("metrics_cache_key: caching is disabled")
+
+func init() {
+	runtime.RegisterResolverInitializer("metrics_cache_key", newMetricsViewCacheKeyResolver)
+}
+
+type metricsViewCacheKeyResolver struct {
+	runtime    *runtime.Runtime
+	instanceID string
+	mvName     string
+	mv         *runtimev1.MetricsViewSpec
+	streaming  bool
+	exectuor   *metricsview.Executor
+	args       *metricsViewCacheKeyResolverArgs
+}
+
+type metricsViewCacheKeyResolverArgs struct {
+	Priority int `mapstructure:"priority"`
+}
+
+type metricsViewCacheKey struct {
+	MetricsView string `mapstructure:"metrics_view"`
+}
+
+func newMetricsViewCacheKeyResolver(ctx context.Context, opts *runtime.ResolverOptions) (runtime.Resolver, error) {
+	tr := &metricsViewCacheKey{}
+	if err := mapstructureutil.WeakDecode(opts.Properties, tr); err != nil {
+		return nil, err
+	}
+
+	args := &metricsViewCacheKeyResolverArgs{}
+	if err := mapstructureutil.WeakDecode(opts.Args, args); err != nil {
+		return nil, err
+	}
+
+	ctrl, err := opts.Runtime.Controller(ctx, opts.InstanceID)
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := ctrl.Get(ctx, &runtimev1.ResourceName{Kind: runtime.ResourceKindMetricsView, Name: tr.MetricsView}, false)
+	if err != nil {
+		return nil, err
+	}
+
+	mv := res.GetMetricsView().State.ValidSpec
+	if mv == nil {
+		return nil, fmt.Errorf("metrics view %q is invalid", res.Meta.Name.Name)
+	}
+
+	security, err := opts.Runtime.ResolveSecurity(opts.InstanceID, opts.Claims, res)
+	if err != nil {
+		return nil, err
+	}
+
+	if !security.CanAccess() {
+		return nil, runtime.ErrForbidden
+	}
+
+	executor, err := metricsview.NewExecutor(ctx, opts.Runtime, opts.InstanceID, mv, res.GetMetricsView().State.Streaming, security, args.Priority)
+	if err != nil {
+		return nil, err
+	}
+
+	return &metricsViewCacheKeyResolver{
+		runtime:    opts.Runtime,
+		instanceID: opts.InstanceID,
+		mvName:     tr.MetricsView,
+		streaming:  res.GetMetricsView().State.Streaming,
+		exectuor:   executor,
+		mv:         mv,
+		args:       args,
+	}, nil
+}
+
+func (r *metricsViewCacheKeyResolver) Close() error {
+	return nil
+}
+
+func (r *metricsViewCacheKeyResolver) Cacheable() bool {
+	return true
+}
+
+func (r *metricsViewCacheKeyResolver) CacheKey(ctx context.Context) ([]byte, bool, error) {
+	var sb strings.Builder
+	sb.WriteString(runtime.ResourceKindMetricsView)
+	sb.WriteString(":")
+	sb.WriteString(r.mvName)
+	sb.WriteString(":")
```

Comment:
If you add the metrics view in `Refs()`, then the caller will take care of adding this to the cache key.